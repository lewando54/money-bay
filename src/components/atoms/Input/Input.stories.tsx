/* eslint-disable @typescript-eslint/consistent-type-assertions */
import React, { useState } from 'react'
import { Meta,StoryObj } from '@storybook/react-native'

import Input from './Input'

interface IInputMockupProps {
    value?: string
    placeholder?: string
    isIconVisible?: boolean
    secureTextEntry?: boolean
    onIconPress?: () => void
}

function StatefulInput({
    value,
    placeholder='Type something',
    isIconVisible=false,
    secureTextEntry=false,
    onIconPress
}: IInputMockupProps){
    const [ textValue, setTextValue ] = useState(value)

    return (
        <Input 
            value={textValue}
            placeholder={placeholder}
            onChange={(newValue) => setTextValue(() => newValue)}
            isIconVisible={isIconVisible}
            secureTextEntry={secureTextEntry}
            onIconPress={onIconPress}
        />
    )
}

const meta = {
    component: StatefulInput,
    argTypes: {
        value: { control: 'text' },
        placeholder: { control: 'text' },
        isIconVisible: { control: 'boolean' },
        secureTextEntry: { control: 'boolean' },
        onIconPress: { action: 'iconPressed' }
    }
} satisfies Meta<typeof StatefulInput>

export default meta
type Story = StoryObj<typeof StatefulInput>

export const Default: Story = {
    args: {
        value: 'Initial value',
        placeholder: 'Type something'
    }
}

export const IconVisible: Story = {
    args: {
        value: 'Initial value',
        placeholder: 'Type something',
        isIconVisible: true
    }
}

export const SecureTextEntry: Story = {
    args: {
        value: 'Initial value',
        placeholder: 'Type something',
        secureTextEntry: true
    }
}
