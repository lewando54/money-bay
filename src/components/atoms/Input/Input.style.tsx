import { StyleSheet } from 'react-native'
import { BODY_TEXT_COLOR } from '../../../styling/GlobalTheme.style'
export default StyleSheet.create({
    container: {
        flexDirection: 'row',
        alignItems: 'center',
        justifyContent: 'space-between',
        borderWidth: 1,
        borderColor: 'white'
    },
    text: {
        maxWidth: '93%'
    },
    focused: {
        borderColor: BODY_TEXT_COLOR,
    }
})
