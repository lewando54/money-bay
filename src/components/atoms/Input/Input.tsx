import React, { useState } from 'react'
import { StyleProp, TextInput, TextStyle, ViewStyle, View, Pressable } from 'react-native'
import inputStyle from './Input.style'
import UniversalContainer from '../UniversalContainer/UniversalContainer'
import { Feather } from '@expo/vector-icons'
import { MAIN_DARK } from '../../../styling/GlobalTheme.style'

interface IInputProps {
    value?: string
    placeholder?: string
    containerStyle?: StyleProp<ViewStyle>
    textStyle?: StyleProp<TextStyle>
    isIconVisible?: boolean
    secureTextEntry?: boolean
    icon?: React.ReactNode
    onChange?: (newValue: string) => void
    onIconPress?: () => void
    testID?: string
}

export default function Input({
    value,
    placeholder='Type something',
    containerStyle,
    textStyle,
    isIconVisible=false,
    secureTextEntry=false,
    icon=<Feather name="eye-off" size={16} color={MAIN_DARK} />,
    onChange,
    onIconPress,
    testID
}: IInputProps){
    const PressableIcon = isIconVisible && <Pressable onPress={onIconPress}>{icon}</Pressable>
    const [isFocused, setIsFocused] = useState(false)

    const onFocus = () => {
        setIsFocused((prev) => !prev)
    }

    const onBlurred = () => {
        setIsFocused((prev) => !prev)
    }

    const handleOnChange = (newValue: string) => {
        if(onChange){
            onChange(newValue)
        }
    }

    return (
        <UniversalContainer 
            padding={14} 
            style={[
                inputStyle.container, 
                containerStyle,
                isFocused && inputStyle.focused
            ]} 
            castShadow={true}>
            <TextInput 
                value={value}
                placeholder={placeholder}
                onChangeText={handleOnChange}
                onFocus={onFocus}
                onBlur={onBlurred}
                testID={testID}
                secureTextEntry={secureTextEntry}
                style={[
                    inputStyle.text, 
                    textStyle, 
                    !isIconVisible && {maxWidth: '100%'}
                ]} 
            />
            {PressableIcon}
        </UniversalContainer>
    )
}
