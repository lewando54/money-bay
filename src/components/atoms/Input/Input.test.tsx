import React from 'react'
import renderer from 'react-test-renderer'
import { fireEvent, render } from '@testing-library/react-native'

import Input from './Input'
import inputStyle from './Input.style'

describe('Input', () => {
    it('should render correctly', () => {
        const tree = renderer.create(<Input />)
        expect(tree).toMatchSnapshot()
    })

    it('should apply focused style when input is focused and remove it when blurred', () => {
        const { getByTestId } = render(<Input testID='textinput-test'/>)

        const input = getByTestId('textinput-test')
        const container = getByTestId('universal-container')
        expect(container.props.style[2]).not.toContain(inputStyle.focused)

        fireEvent(input, 'focus')
        expect(container.props.style[2]).toContain(inputStyle.focused)

        fireEvent(input, 'blur')
        expect(container.props.style[2]).not.toContain(inputStyle.focused)
    })

    it('should pass typed text to passed onChange', () => {
        const onChange = jest.fn()

        const { getByTestId } = render(
            <Input
                value={'text'} 
                placeholder={'placeholder'} 
                secureTextEntry={false} 
                testID='textinput-test'
                onChange={onChange}
            />
        )
        const input = getByTestId('textinput-test')
        fireEvent.changeText(input, 'new value')
        expect(onChange).toBeCalledWith('new value')
    })

    it('should display icon when isIconVisible is true', () => {
        const tree = renderer.create(<Input isIconVisible={true}/>)
        expect(tree).toMatchSnapshot()
    })
})
