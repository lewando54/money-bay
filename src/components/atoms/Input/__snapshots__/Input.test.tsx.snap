// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Input should display icon when isIconVisible is true 1`] = `
<View
  style={
    [
      {
        "backgroundColor": "white",
        "borderRadius": 10,
        "height": "auto",
        "padding": 14,
        "width": "100%",
      },
      {
        "elevation": 20,
        "shadowColor": "#161966",
        "shadowOffset": {
          "height": 6,
          "width": 0,
        },
        "shadowOpacity": 0.08,
        "shadowRadius": 16,
      },
      [
        {
          "alignItems": "center",
          "borderColor": "white",
          "borderWidth": 1,
          "flexDirection": "row",
          "justifyContent": "space-between",
        },
        undefined,
        false,
      ],
    ]
  }
  testID="universal-container"
>
  <TextInput
    onBlur={[Function]}
    onChangeText={[Function]}
    onFocus={[Function]}
    placeholder="Type something"
    secureTextEntry={false}
    style={
      [
        {
          "maxWidth": "93%",
        },
        undefined,
        false,
      ]
    }
  />
  <View
    accessibilityState={
      {
        "busy": undefined,
        "checked": undefined,
        "disabled": undefined,
        "expanded": undefined,
        "selected": undefined,
      }
    }
    accessibilityValue={
      {
        "max": undefined,
        "min": undefined,
        "now": undefined,
        "text": undefined,
      }
    }
    accessible={true}
    collapsable={false}
    focusable={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    testID="input-icon"
  >
    <Text />
  </View>
</View>
`;

exports[`Input should render correctly 1`] = `
<View
  style={
    [
      {
        "backgroundColor": "white",
        "borderRadius": 10,
        "height": "auto",
        "padding": 14,
        "width": "100%",
      },
      {
        "elevation": 20,
        "shadowColor": "#161966",
        "shadowOffset": {
          "height": 6,
          "width": 0,
        },
        "shadowOpacity": 0.08,
        "shadowRadius": 16,
      },
      [
        {
          "alignItems": "center",
          "borderColor": "white",
          "borderWidth": 1,
          "flexDirection": "row",
          "justifyContent": "space-between",
        },
        undefined,
        false,
      ],
    ]
  }
  testID="universal-container"
>
  <TextInput
    onBlur={[Function]}
    onChangeText={[Function]}
    onFocus={[Function]}
    placeholder="Type something"
    secureTextEntry={false}
    style={
      [
        {
          "maxWidth": "93%",
        },
        undefined,
        {
          "maxWidth": "100%",
        },
      ]
    }
  />
</View>
`;
